### ISSUE TITLE
feat(model): general target model batch verification engine with tree-mask and position injection

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12726]
# feat(model): general target model batch verification engine with tree-mask and position injection

```routing
lane: internal/model
paths: internal/model/speculative_verify.go, internal/model/speculative.go
expected_steps: 5
```

## Lane

`internal/model` — inference engine, target model verification, and speculative decoding coordination.

## Problem frame

- Centrality: Core
- Portfolio tier: 2 (serving)
- Priority: P1
- P1: advanced - evaluates multi-token candidate trees in a single target model forward pass instead of serial autoregression
- P2: preserved - guarantees exact output distribution identity and temperature-zero argmax agreement with non-speculative baseline
- P3: advanced - injects decoupled 2D tree attention masks and candidate position IDs into model attention layers
- P4: preserved - coordinates seamlessly with compute tree attention and KV compaction without pipeline stalls

## Parent context

Part of the Native Speculative Acceleration Series (`docs/tickets/native-speculative-acceleration/INDEX.md`), unifying target model verification for both linear chains and branching candidate trees.

## Current state

In `internal/model/speculative.go` and `internal/model/speculative_batch.go`, target model verification coordinates proposal evaluation. However, the existing target forward execution assumes linear candidate chains where candidate token $i$ is assigned position $\text{prefixLen} + i$ and attends causally to preceding candidates. 

When evaluating multi-branch tree proposals (e.g. from EAGLE-3, Medusa, or tree n-grams), candidate tokens at the same depth in different branches share the same positional distance from the prefix, but belong to distinct hypotheses. The current verification loop cannot inject non-sequential position IDs or 2D tree-attention masks into the transformer layers. As a result, candidate trees must currently be flattened into serial checks or restricted to simple linear chains, capping acceptance length at 1.5 to 2.0 tokens.

## Why this is next

Target model forward execution is the primary compute and memory bottleneck during speculative decoding. Evaluating an entire candidate tree of 16 to 32 nodes in a single forward pass maximizes GPU tensor core utilization and arithmetic intensity, amortizing the memory cost of reading the large base model weights over dozens of potential tokens simultaneously.

## Working spine

`SpeculativeEngine` receives `DraftProposal` (linear or tree) -> constructs `TargetVerificationBatch` containing packed tokens, candidate position IDs, and 2D tree mask -> executes target model forward pass evaluating all candidate nodes simultaneously -> extracts logits corresponding to candidate node positions -> runs tree acceptance verification and bonus token selection -> invokes `PruneAndCompactTreeKV` to retain the winning path.

## Core through-line

1. Define `TargetVerificationEngine` and `VerifyProposal(ctx context.Context, session *Session, proposal DraftProposal) (VerificationResult, error)` in `internal/model/speculative_verify.go`.
2. Author batch construction logic that transforms a `CandidateTree` or linear slice into packed query tokens, non-sequential position IDs, and an attention mask conforming to the tree structure.
3. Wire position ID and tree-mask injection into `internal/model/transformer.go`, allowing the attention layer to call `TreeSpecVerifyAttention` (from `internal/compute`) when a tree mask is present.
4. Implement rejection sampling and greedy temperature-zero acceptance oracles over tree topologies: traverse candidate paths from root to leaves, find the longest continuous path matching target argmax (or satisfying nucleus sampling thresholds), and extract one additional bonus token from the terminal accepted node's target logits.
5. Coordinate with `PruneAndCompactTreeKV` (TICKET-02) to immediately compact the KV cache to the accepted sequence upon verification completion.
6. Provide verification tests asserting zero token drift against greedy autoregressive decoding across diverse prompts.

## Gold-plating boundary

Strictly excluded from this work unit:
- Custom sampling algorithms with non-standard temperature schedules or speculative min-p filters.
- Speculative verification over speculative external tool execution (handled by `internal/agentopt`).
- Asynchronous multi-GPU model parallel pipelining of the target verifier.

## Verifiable Witness

```bash
go test -v ./internal/model -run 'TestTargetVerificationEngine' -count=1
```

## Witness

```bash
go test -v ./internal/model -run 'TestTargetVerificationEngine' -count=1
```

### File:Line Seam

- `internal/model/speculative.go:20 (ProposalGenerator)`
- `internal/model/speculative_verify.go:1 (TargetVerificationEngine)`
- `internal/model/speculative_batch.go:10 (SpeculativeBatchConfig)`
- `internal/model/kv_tree_compaction.go:1 (PruneAndCompactTreeKV)`

### Blast Radius & Affected Lanes

- **Primary Lane:** `internal/model`.
- **Unaffected Lanes:** `internal/compute`, `internal/agent`, `platform/*` remain unaffected.

### Quarantined Fallback Mechanism

If tree verification detects non-finite logits, NaN outputs, or tree cycle anomalies, the verifier safely trips the circuit breaker, rejects the entire draft proposal, and falls back to a standard single-token greedy forward step `[SW-VERIFIED]`.

## Scoped Acceptance Criteria

- [ ] 1. [SW-VERIFIED] Implement `TargetVerificationEngine` in `internal/model/speculative_verify.go` supporting packed candidate tokens, arbitrary position IDs, and 2D tree masks.
- [ ] 2. [SW-VERIFIED] Implement tree traversal acceptance logic yielding the longest valid candidate path plus the terminal bonus token.
- [ ] 3. [SW-VERIFIED] Unit tests verify exact sequence identity with greedy single-token generation across 100 benchmark sequence generations.
- [ ] 4. [SW-VERIFIED] Verify seamless handoff to `PruneAndCompactTreeKV` with zero KV fragmentation or slot leakage.
- [ ] 5. [HW-WITNESSED] Verify on physical hardware that single-pass verification of a 16-node candidate tree achieves >= 2.2x token generation speedup over autoregressive decoding.

## Done condition

- [ ] 1. [SW-VERIFIED] Implement `TargetVerificationEngine` in `internal/model/speculative_verify.go` supporting packed candidate tokens, arbitrary position IDs, and 2D tree masks.
- [ ] 2. [SW-VERIFIED] Implement tree traversal acceptance logic yielding the longest valid candidate path plus the terminal bonus token.
- [ ] 3. [SW-VERIFIED] Unit tests verify exact sequence identity with greedy single-token generation across 100 benchmark sequence generations.
- [ ] 4. [SW-VERIFIED] Verify seamless handoff to `PruneAndCompactTreeKV` with zero KV fragmentation or slot leakage.
- [ ] 5. [HW-WITNESSED] Verify on physical hardware that single-pass verification of a 16-node candidate tree achieves >= 2.2x token generation speedup over autoregressive decoding.

## Acceptance gate

`go test -v ./internal/model -run 'TestTargetVerificationEngine' -count=1` exits 0 with 0 boundary violations.

## Closure binding

Resolving commit carries `(fak model)` trailer.

## Likely files

- `internal/model/speculative_verify.go`
- `internal/model/speculative_verify_test.go`
- `internal/model/speculative.go`
- `internal/model/speculative_batch.go`
- `internal/model/kv_tree_compaction.go`

## Expected steps

5
